File: implement-shell-tools/ls/ls.js
const fs = require("fs");
const { allowedNodeEnvironmentFlags } = require("process");

//
// ===== ls Procedure =====
function ls(args) {
  const userArgs = args.slice(2);

  executeFlags(getFlags(userArgs));

  const currentWorkingDirectory = getCurrentWorkingDirectory(args);
  const fsItems = checkArgsLength(getFsItems(userArgs));
  const dirArgs = getDirArgs(currentWorkingDirectory, fsItems);
  const fileArgs = getFileArgs(currentWorkingDirectory, fsItems);

  print(populateOutput(currentWorkingDirectory, fsItems, dirArgs, fileArgs));
}

//
// ===== Flag Handling =====
const flagHandlers = {
  1: Flag1,
  a: Flaga,
};

let printInList = false;
let showAll = false;

function getFlags(userArgs) {
  return userArgs
    .filter((arg) => arg.startsWith("-"))
    .map((arg) => arg.slice(1))
    .join("");
}

function executeFlags(flags) {
  for (const flag of flags) {
    if (flagHandlers[flag]) flagHandlers[flag]();
    else invalidFlagError();
  }
}

function invalidFlagError() {
  console.error(
    `ls: invalid option -- ${flag}\nusage: ls [-@ABCFGHILOPRSTUWXabcdefghiklmnopqrstuvwxy1%,] [--color=when] [-D format] [file ...]`,
  );
  process.exit(1);
}

function Flag1() {
  printInList = true;
}

function makeList(outputString) {
  return outputString.replaceAll("\t", "\n").replaceAll("\n\n", "\n");
}

function Flaga() {
  showAll = true;
}

//
// ===== Extracting Data From Arguments =====
function getCurrentWorkingDirectory(args) {
  const index = args[1].lastIndexOf("/");
  return args[1].slice(0, index + 1);
}

function getFsItems(userArgs) {
  return userArgs.filter((arg) => !arg.startsWith("-"));
}

function checkArgsLength(fsItems) {
  if (fsItems.length == 0) fsItems.push(".");
  return fsItems;
}

function getDirArgs(currentWorkingDirectory, fsItems) {
  return fsItems.filter((p) => fs.statSync(currentWorkingDirectory + p).isDirectory());
}

function getFileArgs(currentWorkingDirectory, fsItems) {
  const fileArgs = fsItems.filter((p) => !fs.statSync(currentWorkingDirectory + p).isDirectory());
  if (showAll == false) removeDotFiles(fileArgs);
  return fileArgs;
}

//
// ===== Populating and Outputting Data =====
function populateOutput(currentWorkingDirectory, fsItems, dirArgs, fileArgs) {
  let outputString = "";
  if (fsItems.length == 1) {
    fileArgs.forEach((file) => {
      outputString += file + " ";
    });
    dirArgs.forEach((dir) => {
      outputString += dirOutput(dir, currentWorkingDirectory);
    });
  } else {
    fileArgs.forEach((file) => {
      outputString += file + "\t";
    });
    dirArgs.forEach((dir) => {
      outputString += "\n\n" + dir + ":\n" + dirOutput(dir, currentWorkingDirectory);
    });
  }
  return outputString;
}

function dirOutput(dir, currentWorkingDirectory) {
  let contents = fs.readdirSync(currentWorkingDirectory + dir);
  let outputStr = "";
  if (showAll) outputStr += ".\t..\t";
  else {
    for (let i = contents.length - 1; i >= 0; i--) {
      contents = removeDotFiles(contents);
    }
  }
  contents.forEach((item) => {
    outputStr += item + "\t";
  });
  return outputStr;
}

function removeDotFiles(fileList) {
  for (let i = fileList.length - 1; i >= 0; i--) {
    if (fileList[i][0] == ".") fileList.splice(i, 1);
  }
  return fileList;
}

function print(outputString) {
  let output = outputString;
  if (printInList) output = makeList(output);
  console.log(output.trimEnd());
}

//
// ===== Run ls =====
ls(process.argv);

File: implement-shell-tools/ls/ls.py
import argparse
import os
from pathlib import Path

# Argument Handling
parser = argparse.ArgumentParser(
    prog="ls",
    description="Print line, word, and byte counts for each file.",
)

parser.add_argument("-1", action="store_true", help="Show output on separate lines")
parser.add_argument("-a", action="store_true", help="Show hidden files")
parser.add_argument("file_system_items", nargs="*", help="Files or folders to display")

args = parser.parse_args()


# ===== ls Procedure =====
def ls(args):
    flag_status = {"print_in_list": False, "show_all": False}
    execute_flags(flag_status)

    fs_items = check_args_length(args.file_system_items)
    dir_args = get_dir_args(fs_items)
    file_args = get_file_args(fs_items, flag_status)

    print_output(
        populate_output(fs_items, dir_args, file_args, flag_status), flag_status
    )


# ===== Flag Handling =====
def execute_flags(flag_status):
    if getattr(args, "1"):
        flag_status["print_in_list"] = True
    if args.a:
        flag_status["show_all"] = True


def make_list(output_string):
    return output_string.replace("\t", "\n").replace("\n\n", "\n")


# Extracting Data from Arguments
def check_args_length(fs_items):
    if len(fs_items) == 0:
        fs_items.append(".")
    return fs_items


def get_dir_args(fs_items):
    return [p for p in fs_items if Path(p).is_dir()]


def get_file_args(fs_items, flag_status):
    file_args = [p for p in fs_items if not Path(p).is_dir()]
    if not flag_status["show_all"]:
        file_args = remove_dot_files(file_args)
    return file_args


# Populating and Outputting Data
def populate_output(fs_items, dir_args, file_args, flag_status):
    output_string = ""

    if len(fs_items) == 1:
        for file in file_args:
            output_string += file + " "
        for dir_item in dir_args:
            output_string += dir_output(dir_item, flag_status)
    else:
        for file in file_args:
            output_string += file + "\t"
        for dir_item in dir_args:
            output_string += f"\n\n{dir_item}:\n" + dir_output(dir_item, flag_status)
    return output_string


def dir_output(dir_item, flag_status):
    target_path = Path.cwd() / dir_item
    contents = os.listdir(target_path)
    output_str = ""
    if flag_status["show_all"]:
        output_str += ".\t..\t"
    else:
        contents = remove_dot_files(contents)

    for item in contents:
        output_str += item + "\t"
    return output_str


def remove_dot_files(file_list):
    return [item for item in file_list if not item.startswith(".")]


def print_output(output_string, flag_status):
    output = output_string
    if flag_status["print_in_list"]:
        output = make_list(output)
    print(output.rstrip())


# ===== Run ls =====
ls(parser.parse_args())
